processes/chicagoISAC.py

```diff
@@ -0,0 +1,114 @@
+import json
+import os
+import requests
+from requests.exceptions import HTTPError, ConnectionError
+
+from .core import CoreProcess
+from mappings.core import MappingError
+from mappings.chicagoISAC import ChicagoISACMapping
+from managers import WebpubManifest
+from logger import createLog
+
+logger = createLog(__name__)
+
+class ChicagoISACProcess(CoreProcess):
+
+    def __init__(self, *args):
+        super(ChicagoISACProcess, self).__init__(*args[:4])
+
+        self.ingestOffset = int(args[5] or 0)
+        self.ingestLimit = (int(args[4]) + self.ingestOffset) if args[4] else 5000
+        self.fullImport = self.process == 'complete' 
+
+        # Connect to database
+        self.generateEngine()
+        self.createSession()
+
+        # S3 Configuration
+        self.s3Bucket = os.environ['FILE_BUCKET']
+        self.createS3Client()
+
+    def runProcess(self):
+        with open('chicagoISAC_metadata.json') as f:
+                chicagoISACData = json.load(f)
+
+        for dict in chicagoISACData:
+            self.processChicagoISACRecord(dict)
+
+        self.saveRecords()
+        self.commitChanges()
+
+    def processChicagoISACRecord(self, record):
+        try:
+            chicagoISACRec = ChicagoISACMapping(record)
+            chicagoISACRec.applyMapping()
+            self.storePDFManifest(chicagoISACRec.record)
+            self.addDCDWToUpdateList(chicagoISACRec)
+            
+        except (MappingError, HTTPError, ConnectionError, IndexError, TypeError) as e:
+            logger.debug(e)
+            logger.warn(ChicagoISACError('Unable to process ISAC record'))
+            
+
+    def storePDFManifest(self, record):
+        for link in record.has_part:
+            itemNo, uri, source, mediaType, flags = link[0].split('|')
+
+            if mediaType == 'application/pdf':
+                recordID = record.identifiers[0].split('|')[0]
+
+                manifestPath = 'manifests/{}/{}.json'.format(source, recordID)
+                manifestURI = 'https://{}.s3.amazonaws.com/{}'.format(
+                    self.s3Bucket, manifestPath
+                )
+
+                manifestJSON = self.generateManifest(record, uri, manifestURI)
+
+                self.createManifestInS3(manifestPath, manifestJSON)
+
+                linkString = '|'.join([
+                    itemNo,
+                    manifestURI,
+                    source,
+                    'application/webpub+json',
+                    flags
+                ])
+                record.has_part.insert(0, linkString)
+                for i in range(1, len(record.has_part)):
+                    if len(record.has_part[i]) > 1:
+                        urlArray = record.has_part[i]
+                        record.has_part.pop(i)
```

Comment:
hmm modifying an array while iterating over its indices could be problematic.  You might need to build a list of indices to delete, then after the loop is finished, delete all the elements in another loop (probably iterating backwards?)

Reply:
I don't believe this will be problematic since before typing this code the mapping for a ISAC record with multiple incoming urls would be [manifestURL, [pdfURL, pdfURL2]] when we want it to be like this [manifestURL, [pdfURL, pdfURL2]. The only element that would be popped would be the second element of the has_part array which will always be an array of pdfURLS and then we would append those elements of that array of pdfURLS back into the has_part array as strings.

Reply:
I see! I think it would be really helpful to add this as a comment explaining the loop logic!